Antiforgery requests topic update

guardrex opened the pull request:
Fixes #5554 

[Internal Review Topic (link will be added when build completes)](#)

* Reorganizes the content a bit
* The usual UE pass attention
* Addressed the comments from #5551 and #5525 
* Promotes the sample to 2.0

Question: There was a (random? lol) mix of "CSRF" and "XSRF" acronyms in the topic. I went with "CSRF," but would you prefer "XSRF" instead?

Rick-Anderson commented:
@blowdart  can you review this or assign it to someone else?

blowdart commented on aspnetcore/security/anti-request-forgery.md:
"simple"? Remove that word. Clearing browser cookies is not simple for the majority of users.

blowdart commented on aspnetcore/security/anti-request-forgery.md:
Session is optional, if we mean session in its classical sense, which seems to be the case here. Cookie auth persists the entire identity in the cookie, it doesn't need sessions. Wording here is confusing.

guardrex replied:
I feel like we could delete the paragraph in this context because the topic pertains to "antirequest forgery," while this paragraph is just providing general background info.

blowdart commented on aspnetcore/security/anti-request-forgery.md:
Again a potential mix up with "session". Also tokens do not have contain information about the user, it could be a reference token which points the application to some user state within itself (probably in a ... err ... session)

guardrex replied:
Yes, this was all held over from the original topic. https://docs.microsoft.com/en-us/aspnet/core/security/anti-request-forgery#cookie-based-authentication

> tokens do not have contain information about the user, it could be a reference token which points the application to some user state within itself (probably in a ... err ... session)

I'll take a stab at updating it along those lines.

blowdart commented on aspnetcore/security/anti-request-forgery.md:
Choose another domain. cloudapp.net (and azurewebsites.net) are isolated in Chrome because I added it to a Chrome blacklist ages ago :)

blowdart commented on aspnetcore/security/anti-request-forgery.md:
Make a note here that you can prevent this by not sharing subdomains with other people - register your own domain.

blowdart commented on aspnetcore/security/anti-request-forgery.md:
Make a note that the HtmlHelper always added the CSRF field. I believe this is the case @NTaylorMullen can confirm.

NTaylorMullen replied:
HtmlHelper's BeginForm generates antiforgery by default if the forms `method` is not `get`

guardrex commented:
@blowdart Updates made.

On the cookie auth "session" para, I just deleted it. I guess Steve just wanted some background basics. I don't think we need that.

On the token remarks, I deleted the first sentence and attempted to include a blurb on how the token might be a reference token.

guardrex commented:
@Rick-Anderson @scottaddie @rachelappel When anyone gets free, this one is ready for a look :eyes:.

Rick-Anderson commented on aspnetcore/security/anti-request-forgery.md:
>o send the token on subsequent requests, either store the token in the browser's local storage or in a cookie. Don't be concerned about CSRF vulnerability if the token is stored in the browser's local storage. CSRF is a concern when the token is stored in a cookie.

Why are you telling them to store it in a cookie only to say that's not secure?

Rick-Anderson commented on aspnetcore/security/anti-request-forgery.md:
~The ASP.NET Core runtime provides some mitigation in that the username is embedded into the field token. Even if a malicious subdomain is able to overwrite a session token, it can't generate a valid field token for the user.~

Why tell them it's OK when it's not?

When hosted in such an environment, the built-in anti-CSRF routines ~still~ can't defend against session hijacking or login CSRF attacks. 

>Shared hosting environments are vulnerable to session hijacking, login CSRF, and other attacks.
I'd lead with that.

scottaddie commented on aspnetcore/host-and-deploy/iis/index.md:
```diff
@@ -277,7 +277,7 @@ If the key ring is stored in memory when the app restarts:
 * All cookie-based authentication tokens are invalidated. 
 * Users are required to sign in again on their next request. 
-* Any data protected with the key ring can no longer be decrypted. This may include [CSRF tokens](xref:security/anti-request-forgery#how-does-aspnet-core-mvc-address-csrf) and [ASP.NET Core MVC tempdata cookies](xref:fundamentals/app-state#tempdata).
+* Any data protected with the key ring can no longer be decrypted. This may include [CSRF tokens](xref:security/anti-request-forgery#aspnet-core-antiforgery-configuration) and [ASP.NET Core MVC tempdata cookies](xref:fundamentals/app-state#tempdata).
```
tempdata --> TempData

scottaddie commented on aspnetcore/security/anti-request-forgery.md:
that changes state --> that change state

guardrex replied:
I split the blame with @Rick-Anderson on that one ... I copied his suggestion word-for-word.

Darn IT, Rick!!!! :smile: lol

scottaddie commented on aspnetcore/security/anti-request-forgery.md:
Should this be generalized to say "ASP.NET Core 2.0"? I assume this statement holds true for Razor Pages apps.

scottaddie commented on aspnetcore/security/anti-request-forgery.md:
Delete the extra space after the opening paren

scottaddie commented on aspnetcore/security/anti-request-forgery.md:
What about forms in Razor Pages templates?

scottaddie commented on aspnetcore/security/anti-request-forgery.md:
Code-fence these option names

guardrex replied:
I have an even better idear. See the next commit in a few minutes.

scottaddie commented on aspnetcore/security/anti-request-forgery.md:
```diff
@@ -185,175 +283,120 @@ public class ManageController : Controller
+### AngularJS
 
-Once a user has authenticated using their username and password, they're issued a token that can be used to identify them and validate that they have been authenticated. The token is stored as a cookie that accompanies every request the client makes. Generating and validating this cookie is done by the cookie authentication middleware. ASP.NET Core provides cookie [middleware](xref:fundamentals/m ...
+AngularJS uses a convention to address CSRF. If the server sends a cookie with the name `XSRF-TOKEN`, the Angular `$http` service adds the cookie value to a header when it sends a request to the server. This process is automatic. The header doesn't need to be set explicitly. The header name is `X-XSRF-TOKEN`. The server should detect this header and validate its contents.
```
It sounds like you're talking about AngularJS here, not Angular. If so, I recommend the following change:

the Angular `$http` --> the AngularJS `$http`

Even better would be to not mention AngularJS at all. It's old, and most of its users have upgraded to Angular.

guardrex replied:
Had to open a separate issue for that update. The topic will have both. https://github.com/aspnet/Docs/issues/5743

scottaddie commented on aspnetcore/security/anti-request-forgery/sample/MvcSample/MvcSample.csproj:
```diff
@@ -1,22 +1,11 @@
 
   <PropertyGroup>
-    <TargetFramework>netcoreapp1.0</TargetFramework>
+    <TargetFramework>netcoreapp2.0</TargetFramework>
```
Consider upgrading to 2.1.

guardrex replied:
It's on the list. I didn't mark myself for it; but if I have time, I'll do it. https://github.com/aspnet/Docs/issues/5495

guardrex commented:
@scottaddie Feedback addressed. I linked those table entries to API browser endpoints. In the process, I noted the ones that are depreciated.
